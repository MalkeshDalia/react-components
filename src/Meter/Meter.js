import React from 'react';
import { props, t, skinnable } from '../utils';
import { warn } from '../utils/log';
import cx from '../utils/classnames';
import FlexView from 'react-flexview';
import find from 'lodash/find';
import every from 'lodash/every';
import isEqual from 'lodash/isEqual';
import sortBy from 'lodash/sortBy';

const Range = t.refinement(t.struct({
  startValue: t.Number,
  endValue: t.Number,
  fillingColor: t.maybe(t.String),
  labelColor: t.maybe(t.String),
  backgroundColor: t.maybe(t.String)
}), r => r.startValue < r.endValue, 'Range');

const Ranges = t.refinement(t.list(Range), (rangeList) => {

  const isOverlapped = (range1, range2) => (
    Math.max(range1.startValue, range2.startValue) < Math.min(range1.endValue, range2.endValue)
  );

  const noOverlappingRanges = ranges => {
    return every(ranges, (range1, i) => {
      return every(ranges.slice(0, i).concat(ranges.slice(i + 1)), range2 => {
        return !isOverlapped(range1, range2);
      });
    });
  };

  return noOverlappingRanges(rangeList);
}, 'Ranges');

export const Props = t.refinement(t.struct({
  value: t.Number,
  min: t.maybe(t.Number),
  max: t.maybe(t.Number),
  labelFormatter: t.maybe(t.Function),
  ranges: t.maybe(Ranges),
  baseLabelColor: t.maybe(t.String),
  baseFillingColor: t.maybe(t.String),
  baseBackgroundColor: t.maybe(t.String),
  id: t.maybe(t.String),
  className: t.maybe(t.String),
  style: t.maybe(t.Object)
}), ({ min, max }) => min < max, 'Props');

const isFullyFilled = (ranges, min, max) => {
  if (!ranges) {
    return false;
  }
  const comparableRanges = ranges.concat({ startValue: max, endValue: min });
  const sortedStartValueList = sortBy(comparableRanges, 'startValue').map(range => range.startValue);
  const sortedEndValueList = sortBy(comparableRanges, 'endValue').map(range => range.endValue);
  return isEqual(sortedStartValueList, sortedEndValueList);
};

const computePercentage = (value, min, max) => (
  Math.abs((value - min) * 100 / (max - min))
);

const labelFormatter = (value, min, max) => (
  `${computePercentage(value, min, max)}%`
);

/**
 * A Meter is a simple UI component used to display a measurement (usually a percentage) on a known scale.
 * @param value - current value
 * @param min - minimum value
 * @param max - maximum value
 * @param labelFormatter - function in which you can define a custom label format
 * @param ranges - array of Object in which you can define startValue, endValue, labelColor, fillingColor
 * @param baseLabelColor - fallback labelColor
 * @param baseFillingColor - fallback fillingColor
 * @param baseBackgroundColor - fallback backgroundColor
 */
@skinnable()
@props(Props)
export default class Meter extends React.Component {

  static defaultProps = {
    min: 0,
    max: 100,
    labelFormatter
  };

  componentDidMount() {
    this.logWarnings();
  }

  logWarnings = () => {
    const {
      ranges,
      max,
      min,
      baseFillingColor,
      baseBackgroundColor
    } = this.props;
    warn(() => {
      if (isFullyFilled(ranges, min, max) && (baseFillingColor || baseBackgroundColor)) {
        return 'baseFillingColor or baseBackgroundColor is not needed, ranges are fully filled';
      }
      return undefined;
    });
    warn(() => {
      if (!(isFullyFilled(ranges, min, max) || (baseFillingColor && baseBackgroundColor))) {
        return 'You should pass both baseFillingColor and baseBackgroundColor, ranges are not fully filled';
      }
      return undefined;
    });
  };

  computeStyles = () => {
    const {
      value,
      min,
      max,
      ranges,
      baseFillingColor,
      baseLabelColor,
      baseBackgroundColor
    } = this.props;

    const range = find(ranges, ({ startValue, endValue }) => (
      value >= startValue && value <= endValue
    ));
    return {
      basisSize: `${computePercentage(value, min, max)}%`,
      fillingStyle: {
        backgroundColor: range ? range.fillingColor : baseFillingColor
      },
      labelStyle: {
        color: range ? range.labelColor : baseLabelColor
      },
      barStyle: {
        backgroundColor: range ? range.backgroundColor : baseBackgroundColor
      }
    };
  }

  getLocals() {
    const {
      className,
      labelFormatter,
      ...props
    } = this.props;

    const styles = this.computeStyles();

    return {
      ...props,
      className: cx('meter', className),
      fillingStyle: styles.fillingStyle,
      labelStyle: styles.labelStyle,
      barStyle: styles.barStyle,
      basisSize: styles.basisSize,
      formattedLabel: labelFormatter(props.value, props.min, props.max)
    };
  }

  template({ id, className, style, fillingStyle, labelStyle, barStyle, basisSize, formattedLabel }) {
    return (
      <FlexView {...{ id, className, style }} grow>
        <FlexView
          className='bar'
          grow
          style={barStyle}
        >
          <FlexView
            className='filling'
            basis={basisSize}
            style={fillingStyle}
          />
        </FlexView>
        <FlexView
          className='label'
          shrink={false}
          style={labelStyle}
        >
          {formattedLabel}
        </FlexView>
      </FlexView>
    );
  }
}
