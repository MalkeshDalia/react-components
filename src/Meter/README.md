# Meter

A Meter is a simple UI component used to display a measurement (usually a percentage) on a known scale.

## Props
|Name|Type|Default|Description|
|----|----|-------|-----------|
| **value** | <code>Number</code> |  | **required**. Current value |
| **min** | <code>Number</code> | <code>0</code> | *optional*. Minimum value |
| **max** | <code>Number</code> | <code>100</code> | *optional*. Maximum value |
| **labelFormatter** | <code>Function</code> | <code>"labelFormatter"</code> | *optional*. Function in which you can define a custom label format |
| **ranges** | <code>Ranges</code> |  | *optional*. Array of Object in which you can define startValue, endValue, labelColor, fillingColor |
| **baseLabelColor** | <code>String</code> |  | *optional*. Fallback labelColor |
| **baseFillingColor** | <code>String</code> |  | *optional*. Fallback fillingColor |
| **baseBackgroundColor** | <code>String</code> |  | *optional*. Fallback backgroundColor |
| **id** | <code>String</code> |  | *optional*. Custom `id` for wrapper element |
| **className** | <code>String</code> |  | *optional*. Additional `className` for wrapper element |
| **style** | <code>Object</code> |  | *optional*. Inline-style overrides for wrapper element |