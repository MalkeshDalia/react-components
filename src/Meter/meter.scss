@import '../utils/theme/variables.scss';

.meter {
  .bar {
    overflow: hidden;
    border-radius: 10px;
    height: 18px;
    background-color: $brc-paleGrey;
  }

  .filling {
    border-bottom-left-radius: 10px;
    border-top-left-radius: 10px;
  }
}
