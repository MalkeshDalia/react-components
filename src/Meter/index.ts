import Meter, { MeterProps } from './Meter';
export default Meter;
export { MeterProps };
