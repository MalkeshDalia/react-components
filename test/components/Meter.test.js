import React from 'react';
import renderer from 'react-test-renderer';

import Meter from '../../src/Meter';

const exampleProps = {
  ...Meter.defaultProps,
  id: '12345',
  className: 'fancy-class-name',
  style: { margin: 10, position: 'relative' },
  value: 25,
  min: 0,
  max: 50,
  ranges: [
    { startValue: 0, endValue: 50, fillingColor: 'green', labelColor: 'red', backgroundColor: 'grey' },
    { startValue: 50, endValue: 80, fillingColor: 'yellow' },
    { startValue: 80, endValue: 100 }
  ],
  baseLabelColor: '#000000',
  baseFillingColor: '#ccc',
  labelFormatter: () => 'test'
};

const meter = new Meter(exampleProps);

describe('Meter', () => {

  it('renders correctly', () => {
    const component = renderer.create(
      <Meter {...exampleProps} />
    );
    expect(component).toMatchSnapshot();
  });

  describe('getLocals', () => {

    it('computes className', () => {
      const { className } = meter.getLocals();
      expect(className).toContain('meter');
      expect(className).toContain('fancy-class-name');
    });

    it('uses labelFormatter', () => {
      const { formattedLabel } = meter.getLocals();
      expect(formattedLabel).toBe('test');
    });

    it('computes barStyle', () => {
      const { barStyle, ranges } = meter.getLocals();
      expect(barStyle.backgroundColor).toBe(ranges[0].backgroundColor);
    });

    it('computes fillingStyle', () => {
      const { fillingStyle, ranges } = meter.getLocals();
      expect(fillingStyle.backgroundColor).toBe(ranges[0].fillingColor);
    });

    it('computes labelStyle', () => {
      const { labelStyle, ranges } = meter.getLocals();
      expect(labelStyle.color).toBe(ranges[0].labelColor);
    });

    it('computes basisSize', () => {
      const { basisSize } = meter.getLocals();
      expect(basisSize).toBe('50%');
    });

    it('computes basisSize with custom min and max', () => {
      const meter = new Meter({
        ...Meter.defaultProps,
        value: 150,
        min: 100,
        max: 200,
        labelFormatter: () => 'test'
      });
      const { basisSize } = meter.getLocals();
      expect(basisSize).toBe('50%');
    });

    it('computes basisSize for custom negative min', () => {
      const meter = new Meter({
        ...Meter.defaultProps,
        value: 0,
        min: -100,
        max: 100,
        labelFormatter: () => 'test'
      });
      const { basisSize } = meter.getLocals();
      expect(basisSize).toBe('50%');
    });

    it('computes background color for filling', () => {
      const meter = new Meter({
        ...Meter.defaultProps,
        value: 60,
        ranges: [
          { startValue: 50, endValue: 80, fillingColor: 'yellow' }
        ],
        labelFormatter: () => 'test'
      });
      const { fillingStyle, ranges } = meter.getLocals();
      expect(fillingStyle.backgroundColor).toBe(ranges[0].fillingColor);
    });

    it('defaults to the base color as background color if there\'s no matching range', () => {
      const meter = new Meter({
        ...Meter.defaultProps,
        value: 20,
        baseFillingColor: '#ccc',
        ranges: [
          { startValue: 50, endValue: 80, fillingColor: 'yellow' }
        ],
        labelFormatter: () => 'test'
      });
      const { fillingStyle, baseFillingColor } = meter.getLocals();
      expect(fillingStyle.backgroundColor).toBe(baseFillingColor);
    });

    it('doesn\'t define a background color if there\'s no matching range and no default is given', () => {
      const meter = new Meter({
        ...Meter.defaultProps,
        value: 20,
        ranges: [
          { startValue: 50, endValue: 80, fillingColor: 'yellow' }
        ],
        labelFormatter: () => 'test'
      });
      const { fillingStyle } = meter.getLocals();
      expect(fillingStyle.backgroundColor).toBeUndefined();
    });

    it('defaults to base color as bar background color should if there\'s no matching range', () => {
      const meter = new Meter({
        ...Meter.defaultProps,
        value: 20,
        baseBackroundColor: '#ccc',
        ranges: [
          { startValue: 50, endValue: 80, backgroundColor: 'yellow' }
        ],
        labelFormatter: () => 'test'
      });
      const { barStyle, baseBackgroundColor } = meter.getLocals();
      expect(barStyle.backgroundColor).toBe(baseBackgroundColor);
    });

    it('doesn\'t define a bar background color if there\'s no matching range and no default is given', () => {
      const meter = new Meter({
        ...Meter.defaultProps,
        value: 20,
        ranges: [
          { startValue: 50, endValue: 80, backgroundColor: 'yellow' }
        ],
        labelFormatter: () => 'test'
      });
      const { barStyle } = meter.getLocals();
      expect(barStyle.backgroundColor).toBeUndefined();
    });
  });
});
