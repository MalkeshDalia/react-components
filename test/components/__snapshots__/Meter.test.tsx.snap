// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Meter renders correctly 1`] = `
<div
  className="react-flex-view align-content-center meter fancy-class-name"
  id="12345"
  style={
    Object {
      "WebkitFlex": "1 1 auto",
      "flex": "1 1 auto",
      "margin": 10,
      "msFlex": "1 1 auto",
      "position": "relative",
    }
  }
>
  <div
    className="react-flex-view bar"
    style={
      Object {
        "WebkitFlex": "1 1 auto",
        "background": "grey",
        "flex": "1 1 auto",
        "msFlex": "1 1 auto",
      }
    }
  >
    <div
      className="react-flex-view filling"
      style={
        Object {
          "WebkitFlex": "0 0 25%",
          "background": "green",
          "flex": "0 0 25%",
          "msFlex": "0 0 25%",
        }
      }
    />
  </div>
  <div
    className="react-flex-view label"
    style={
      Object {
        "WebkitFlex": "0 0 auto",
        "color": "red",
        "flex": "0 0 auto",
        "msFlex": "0 0 auto",
      }
    }
  >
    test
  </div>
</div>
`;
